packages/aggregator/src/analytics/aggregateAnalytics/adjustOptionsToAggregationList.js

```diff
@@ -0,0 +1,134 @@
+ * This function preserves the existing functionality by retaining the original period if
+ * there was no date range adjustment
+ */
+const getAdjustedPeriod = (originalOptions, adjustedOptions) => {
+  const areDateRangesEqual =
+    originalOptions?.startDate === adjustedOptions?.startDate &&
+    originalOptions?.endDate === adjustedOptions?.endDate;
+
+  return areDateRangesEqual
+    ? originalOptions.period
+    : convertDateRangeToPeriodString(adjustedOptions.startDate, adjustedOptions.endDate);
+};
+
+const getAdjustedDateRange = (dateRange, aggregationList) =>
+  aggregationList.reduce((currentRange, aggregation) => {
+    const getDateRange = AGGREGATION_TYPE_TO_DATE_RANGE_GETTER[aggregation.type];
+    return getDateRange ? getDateRange(currentRange, aggregation.config) : currentRange;
+  }, dateRange);
+
+const shouldFetchDataSourceEntities = ({ config: { dataSourceEntityType } = {} } = {}) =>
+  dataSourceEntityType;
+
+const shouldFetchRelations = ({
+  type,
+  config: { aggregationEntityType, dataSourceEntityType } = {},
+} = {}) => aggregationEntityType && dataSourceEntityType && !(type === AGGREGATION_TYPES.RAW);
+
+const getAdjustedOrganisationUnitsAndAggregations = async (
+  context,
+  fetchOptions,
+  aggregationList,
+) => {
+  const { hierarchy, organisationUnitCodes } = fetchOptions;
+  if (!aggregationList.some(shouldFetchDataSourceEntities)) {
+    return [organisationUnitCodes, aggregationList];
+  }
+
+  const { session } = context;
+
+  let adjustedOrganisationUnitCodes;
+  const adjustedAggregations = [];
+  const entityConnection = new EntityConnection(session);
+  await Promise.all(
+    aggregationList.map(async aggregation => {
+      if (!shouldFetchDataSourceEntities(aggregation)) {
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      if (!adjustedOrganisationUnitCodes && !shouldFetchRelations(aggregation)) {
+        const { dataSourceEntityType, dataSourceEntityFilter } = aggregation.config;
+        adjustedOrganisationUnitCodes = await entityConnection.getDataSourceEntities(
+          hierarchy,
+          organisationUnitCodes,
+          dataSourceEntityType,
+          dataSourceEntityFilter,
+        );
+        adjustedAggregations.push(aggregation);
+        return;
+      }
+
+      const {
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      } = aggregation.config;
+      const [
+        dataSourceEntities,
+        relations,
+      ] = await entityConnection.getDataSourceEntitiesAndRelations(
+        hierarchy,
+        organisationUnitCodes,
+        aggregationEntityType,
+        dataSourceEntityType,
+        dataSourceEntityFilter,
+      );
+
+      if (!adjustedOrganisationUnitCodes) {
+        adjustedOrganisationUnitCodes = dataSourceEntities;
+      }
+      adjustedAggregations.push({
+        ...aggregation,
+        config: { ...aggregation.config, orgUnitMap: relations },
```

Comment:
Great!!! So much simpler for the `orgUnitMap` to be built here!! 👏 👏

Reply:
Definitely much simpler than the code that builds it in web-config-server 😛